File: programming_languages/perl/utf8.pl
use strict;
use Encode;
use open qw( :std :encoding(UTF-8) );
use Unicode::GCString ;

open IN, "../utf8.txt"; # this file is UTF-8

open(my $OUT_1, ">", "output.txt");

sub PrintBoth {
   print $OUT_1 $_[1];
   print $_[1];
}

while ( <IN> ) {
    chomp;
    my $line = $_;
    my $gcstring = Unicode::GCString->new( $line );
    PrintBoth($OUT_1, "String: $line\n");
    my $refType = ref($line);
    PrintBoth($OUT_1, "String Type: ${refType}\n");
    PrintBoth($OUT_1, "List: ");
    foreach my $char (split //, $line) {
      PrintBoth($OUT_1, "$char, ");
    }
    PrintBoth($OUT_1, "\n");
    PrintBoth($OUT_1, "Length: ${\length($line)}\n");
    PrintBoth($OUT_1, "Seventh character: ${\substr($line, 6, 1)}\n");
    my $count = 0;
    PrintBoth($OUT_1, "Grapheme(s): ");
    while ( my $gcchar = $gcstring->next ) {

        $count++;
        PrintBoth($OUT_1, "$gcchar, ");
    }
    
    PrintBoth($OUT_1, "\nGrapheme(s) Length: $count\n");
    PrintBoth($OUT_1, "Seventh Grapheme: ${\$gcstring->substr(6, 1)->as_string}\n");
    PrintBoth($OUT_1, "\n");

# Grapheme(s): S, M, O, N, E, Ć, T, E, N,
# Grapheme(s) Length: 9
}

close IN;
close OUT_1;

File: docs/programming_languages/python2/utf8.py
#! /usr/bin/python2

import codecs
import regex

output = open("output.txt","w")
with codecs.open("../utf8.txt",encoding="utf-8",mode="r") as exampleFile:
    for line in exampleFile:
        encodedLine = line.encode("utf-8")
        # Has new line from file input ensure we write out the proper encoded version
        output.write("String: "+encodedLine)
        print("String: "+encodedLine[:-1])

        output.write("String type: "+str(type(encodedLine))+"\n")
        print("String type: "+str(type(encodedLine)))

        # Note that if you preface a string with u it becomes unicode instead of ascii
        strList = u", ".join(list(line)).encode('utf-8')
        output.write("List: " + strList)
        print("List: " + strList[:-1])

        output.write("Length: "+str(len(encodedLine))+"\n")
        print("Length: "+str(len(encodedLine)))

        output.write("Seventh character: "+ line[6].encode('utf-8')+"\n")
        print("Seventh character: " + line[6].encode('utf-8'))

        graphemeArr = regex.findall(r'\X', line, regex.U)
        graphemes = u", ".join(graphemeArr).encode('utf-8')
        output.write("Grapheme(s): " + graphemes + "\n\n")
        print "Grapheme(s): " + graphemes, 

        output.write("Grapheme(s) Length: " + str(len(graphemeArr)) + "\n\n")
        print("Grapheme(s) Length: " + str(len(graphemeArr)) + "\n")
output.close()
exampleFile.close()

print("This output is also available in output.txt")
